backends/optimized/ceed-opt-basis.c

```diff
@@ -119,13 +152,21 @@ static int CeedBasisApply_Opt(CeedBasis basis, CeedTransposeMode tmode,
           u += nqpt;
         }
       }
+      // Second half of the contractions
+      // CEED_NOTRANSPOSE - Gradient at quadrature points
+      // CEED_TRANSPOSE - Interpolate back to dofs
       P = basis->Q1d, Q = basis->Q1d;
+      outsize = nqpt*dim;
+      midsize = ncomp*Q*CeedPowInt(P>Q?P:Q, dim-1);
       if (tmode == CEED_TRANSPOSE) {
         P = basis->Q1d, Q = basis->P1d;
+        outsize = ncomp*CeedPowInt(basis->P1d, dim);
       }
       pre = ncomp*CeedPowInt(P, dim-1), post = 1;
```

Comment:
So this batching looks good, but we need to change one thing. The number of elements should be the last index (which means element restriction needs to be permuted), so that `post = batchsize`. This will make it so that every operation within TensorContract will operate on at least `batchsize` contiguous values, so that a consistent vectorization strategy applies. (Otherwise vectorization is messier and dependent on the sizes `P` and `Q`.) If that change in (temporary) E-vector data layout makes sense and you can implement it, then I can show how to specialize the tensor contraction so the compiler will generate good code. (It's too generic now for the compiler to vectorize well, but we need the data layout change to have an easy path to good implementation -- or any path to good implementation at some data sizes.)

Reply:
So we would always apply a restriction operation in this case, even for `CEED_RESTRICTION_IDENTITY` to permute the dofs? That makes sense to me, but I am not sure how we would make this permuting compatible with the `QFunctionApply`. Right now qfunctions require contiguous inputs and outputs. Would we interlace for basis actions, separate for qfunctions, then interlace again for the basis before separating for the final basis action? That feels like a lot of copying.

Reply:
The qfunction doesn't care whether the "next" quadrature point is in the same element or a different element. We just need to keep all Q-vectors in the same ordering (e.g., interlaced within batches), but there are no additional implied permutations/copies.

It is true that we'll need a permutation for `CEED_RESTRICTION_IDENTITY` (and perhaps eventually provide a way for that data to be provided in a batched ordering).